### ISSUE TITLE
Type macros may follow a different naming convention

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/api-guidelines#145]
https://github.com/rust-lang-nursery/api-guidelines/issues/135#issuecomment-343698387 suggests a nice one:

> I've been using `PascalCase!` for macros that expand into types. (e.g. `Array![i32; 5; 4; 3]` which expands into `[[[i32; 3]; 4]; 5]`)

I did not consider type macros when adding a case convention for macros in C-CASE. I have also never used a type macro myself. Will need to survey the ecosystem and see whether `PascalCase!` for type macros would be a good recommendation.

**Comment by KodrAus** [comment]
This seems like an amendment we'd want to make now. In the intervening years I think PascalCase has emerged as a standard for type macros (like [`syn::Token!`](https://docs.rs/syn/1.0.55/syn/macro.Token.html)).

**Comment by KodrAus** [comment]
@rfcbot fcp merge

**Comment by KodrAus** [comment]
@rfcbot fcp merge

This proposes amending our naming conventions to recommend type macros use PascalCase naming instead of snake_case.

**Comment by CodesInChaos** [comment]
Are there any type macros in the standard library already?

**Comment by PlasmaPower** [comment]
Yes, there's `vec!` and `format_args!`, both using snake case instead of pascal case.

**Comment by sfackler** [comment]
Both of those macros expand into values, not types.
